File: Week 4/TicTacToe Client/Assets/Scripts/PacketBuilder.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public static class PacketBuilder
{
    public static Buffer Join(string username)
    {
        int packetLength = 5 + username.Length;
        Buffer packet = Buffer.Alloc(packetLength);

        packet.WriteString("JOIN");
        packet.WriteUInt8((byte)username.Length, 4);
        packet.WriteString(username, 5);

        return packet;
    }
    public static Buffer Chat(string message)
    {
        int packetLength = 5 + message.Length;
        Buffer packet = Buffer.Alloc(packetLength);

        packet.WriteString("CHAT");
        packet.WriteUInt8((byte)message.Length, 4);
        packet.WriteString(message, 5);

        return packet;
    }
    public static Buffer Play(int x, int y)
    {
        Buffer packet = Buffer.Alloc(6);

        packet.WriteString("PLAY");
        packet.WriteUInt8((byte)x, 4);
        packet.WriteUInt8((byte)y, 5);

        return packet;
    }
}

File: Week 4/ttt-server/packet-builder.js

exports.PacketBuilder = {
	join(responseID){
		const packet = Buffer.alloc(5);
		
		packet.write("JOIN", 0);
		packet.writeUInt8(responseID, 4);

		return packet;
	},
	chat(){

	},
	update(game){
		const packet = Buffer.alloc(15);
		packet.write("UPDT", 0);
		packet.writeUInt8(game.whoseTurn, 4);
		packet.writeUInt8(game.whoHasWon, 5);

		let offset = 6;
		for(let y = 0; y < game.board.length; y++){
			for(let x = 0; x < game.board[y].length; x++){

				packet.writeUInt8(game.board[y][x], offset);
				offset++;

			}
		}

		return packet;
	},
};
